### PR TITLE
[IMP] web_editor: toggle background shape on mobile

### PR DESCRIPTION
**Pull request by robinlej**
[IMP] web_editor: toggle background shape on mobile

Currently, all background shapes of snippets are hidden on mobile,
because most of them don't look great on small screens. This commit
adds a button in the website editor to toggle the shape on mobile.
It remains hidden by default if added on desktop, but displayed by
default if added on mobile.

task-3096132

### CODE REVIEW

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Seems good, some comments to make it perfect before sending this to testing :+1:

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 1]
```diff
@@ -6883,6 +6884,20 @@ registry.BackgroundShape = SnippetOptionWidget.extend({
...
+     * 
+     * @see this.selectClass for params
+     */
+    showOnMobile(previewMode, widgetValue, params) {
+        this._handlePreviewState(previewMode, () => {
+            const showOnMobile = !this._getShapeData().showOnMobile;
+            const shape = this.$target.find('> .o_we_shape')[0];
+            shape.classList.toggle('o_shape_show_mobile');
+
+            return {showOnMobile};
```
[the remark this answers is not in the document]
Comment (robinlej) [reply addons/web_editor/static/src/js/editor/snippets.options.js:6898]: Not that I can see in `color()` and `_flipShape()`:
- https://github.com/odoo/odoo/blob/947a7ac915a4614885eec105dd23f75b2149f5ef/addons/web_editor/static/src/js/editor/snippets.options.js#L6862-L6868
- https://github.com/odoo/odoo/blob/947a7ac915a4614885eec105dd23f75b2149f5ef/addons/web_editor/static/src/js/editor/snippets.options.js#L6969-L6977

File: addons/web_editor/static/src/scss/web_editor.frontend.scss [hunk 2]
```diff
@@ -18,7 +18,7 @@
 }
 
 @include media-breakpoint-down(md) {
-    .o_we_shape {
+    .o_we_shape:not(.o_shape_show_mobile) {
```
[the remark this answers is not in the document]
Comment (robinlej) [reply addons/web_editor/static/src/scss/web_editor.frontend.scss:21]: I initially did something like that (with `d-none` and variants) but a unique toggle class seemed more in line with the code. I can go back to bootstrap classes if it's better, it's just a few lines inside the showOnMobile method.

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (robinlej) [reply addons/web_editor/static/src/js/editor/snippets.options.js:6898]: Maybe I read it wrong, but `_handlePreviewState` passes what its function parameter returns to `_markShape`, which in turn recreates a full object from `_getShapeData()` before modifying it with the parameter. So it seems to be the expected use.
https://github.com/odoo/odoo/blob/947a7ac915a4614885eec105dd23f75b2149f5ef/addons/web_editor/static/src/js/editor/snippets.options.js#L7106

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
@robinlej There is 3 commits in your PR, I don't think this is what you wanted me to review? :wink: 

image

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 3]
```diff
@@ -7065,6 +7079,10 @@ registry.BackgroundShape = SnippetOptionWidget.extend({
             $(shapeContainer).css('background-image', '');
             $(shapeContainer).css('background-position', '');
         }
+        shapeContainer.classList.remove('o_shape_show_mobile');
+        if (showOnMobile) {
+            shapeContainer.classList.add('o_shape_show_mobile');
+        }
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:7085]: Make this a one-liner thanks to `classList.toggle`

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
@robinlej A very small comment. Once adapted set the task in RD Dev - Green so this can go through testing :+1: Don't forget it should already been moved to RD Dev as soon as you re-started working on it, I see it is still in RD Review :wink:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
@pierre-pde @robinlej Better, you can now add a shape on desktop, make it visible in mobile, then select a different shape by previsualizing them in mobile... but when you start in mobile, it does not work:
- Enter edit mode, add a snippet
- Switch to mobile view
- Add a shape: the selection stuff opens but previsualizing does not work. Sure the user can click on "show in mobile" then reopen the dropdown but... strange that dropdown opened then. I think that when you are in mobile and you add a shape, that checkbox to show in mobile should be enabled by default.

Also @robinlej when changing: also rebase this branch + remove the "CSS updated" in the commit message. Explain the what and the why, not the how in commit messages (unless very technical stuff). The how is the diff you made: if the diff is not clear enough, there should be explanation **as comments in the code** :wink: Actually, this CSS rule `&:not(.o_we_so_color_palette) ~ we-button:not(:hover):last-child {` could be commented as it is not obvious what it does (I may myself come back to it in one year and not having a clue of what it is without blaming) :+1:

**Comment by pierre-pde** [comment]
Hi @qsm-odoo , good point. I forgot to think about editing in mobile view...
@robinlej , 100% worth implementing the suggestion of Quentin.
If a shape is added on mobile view, enable the "show on mobile" automatically :)

**Review by qsm-odoo — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**